unpack-trees.c

```diff
@@ -1741,6 +1741,51 @@ int unpack_trees(unsigned len, struct tree_desc *t, struct unpack_trees_options
 		ensure_full_index(o->dst_index);
 	}
 
+	/*
+	 * If the prefix is or is contained within a sparse directory, the index
+	 * needs to be expanded to traverse with the specified prefix. Note that
+	 * only the src_index is checked because the prefix is only specified
+	 * in cases where src_index == dst_index.
+	 */
+	if (o->prefix && o->src_index->sparse_index) {
+		int i, ce_len;
+		struct cache_entry *ce;
+		int prefix_len = strlen(o->prefix);
+
+		if (prefix_len > 0) {
+			for (i = 0; i < o->src_index->cache_nr; i++) {
+				ce = o->src_index->cache[i];
+				ce_len = ce_namelen(ce);
+
+				if (!S_ISSPARSEDIR(ce->ce_mode))
+					continue;
+
+				/*
+				 * Normalize comparison length for cache entry vs. prefix -
+				 * either may have a trailing slash, which we do not want to
+				 * compare (can assume both are directories).
+				 */
+				if (ce->name[ce_len - 1] == '/')
+					ce_len--;
+				if (o->prefix[prefix_len - 1] == '/')
+					prefix_len--;
```

Comment:
Oddly, we _want_ the paths to end with a slash, because "A/Dir" is a prefix of "A/Directory" but "A/Dir/" is not a prefix of "A/Directory/". One way to check for this is actually to check that `ce_len == prefix_len || o->prefix[ce_len] == '/'` in the same condition as the `!strcmp()`.

Reply:
I went with an `if (opposite of that condition) continue;` block (to mirror the style of the rest of the block, but otherwise this should be fixed. Good catch!

Reply:
DeMorgan to the rescue!